C error found while compiling - fixed array with size defined by const

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 2ab1523
OS: linux, "Arch Linux"
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz

vexe: /usr/lib/vlang/v
vexe mtime: 2024-11-06 10:58:55

vroot: NOT writable, value: /usr/lib/vlang
VMODULES: OK, value: /home/onraj/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.47.0
Git vroot status: Error: fatal: not a git repository (or any of the parent directories): .git
.git/config present: false

CC version: cc (GCC) 14.2.1 20240910
thirdparty/tcc status: Error: fatal: detected dubious ownership in repository at '/usr/lib/vlang/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/lib/vlang/thirdparty/tcc
 Error: fatal: detected dubious ownership in repository at '/usr/lib/vlang/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/lib/vlang/thirdparty/tcc
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg src/ssh.v`

```v
module main

import crypto.ed25519

struct SSHClientHello {
	version u8
	public_key [ed25519.public_key_size]u8
}

struct SSHServerHello {
	version u8
	send_me [ed25519.public_key_size]u8 // Randomly encrypted string which will be decrypted by the client and the result will be send back to the server as authentication
}

struct SSHClientProof {
	result [ed25519.public_key_size]u8
}

fn new_client_hello(public_key [ed25519.public_key_size]u8) {
}
```

**What did you expect to see?**

The file should compile

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/ssh.01JC6CBF40EHSCDN2PVZY3B7F2.tmp.c:3935: error: ',' expected (got "public_key")
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzJlNDc5MGU3Y2M1YTc4NTQyOTkyYTkiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.QoAcKLtiBP2DUSoI60Bgmc8txMExcf78EiiR8sKURoY">Huly&reg;: <b>V_0.6-21257</b></a></sub>

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -272,6 +272,14 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 				c.error('result type arguments are not supported', param.type_pos)
 			}
 			arg_typ_sym := c.table.sym(param.typ)
+			// resolve unresolved fixed array size e.g. [mod.const]array_type
+			if arg_typ_sym.info is ast.ArrayFixed
+				&& c.array_fixed_has_unresolved_size(arg_typ_sym.info) {
+				mut size_expr := unsafe { arg_typ_sym.info.size_expr }
+				param.typ = c.eval_array_fixed_sizes(mut size_expr, 0, arg_typ_sym.info.elem_type)
+				mut v := node.scope.find_var(param.name) or { continue }
+				v.typ = param.typ
+			}
 			if arg_typ_sym.info is ast.Struct {
 				if !param.typ.is_ptr() && arg_typ_sym.info.is_heap { // set auto_heap to promote value parameter
 					mut v := node.scope.find_var(param.name) or { continue }

--- a/vlib/v/tests/fns/param_fixed_array_const_test.v
+++ b/vlib/v/tests/fns/param_fixed_array_const_test.v
@@ -2,20 +2,6 @@ module main
 
 import crypto.ed25519
 
-struct SSHClientHello {
-	version    u8
-	public_key [crypto.ed25519.public_key_size]u8
-}
-
-struct SSHServerHello {
-	version u8
-	send_me [crypto.ed25519.public_key_size]u8
-}
-
-struct SSHClientProof {
-	result [crypto.ed25519.public_key_size]u8
-}
-
 fn new_client_hello(public_key [ed25519.public_key_size]u8) {
 	assert public_key.len == ed25519.public_key_size
 	assert public_key.len != 0

--- a/vlib/v/tests/fns/param_fixed_array_const_test.v
+++ b/vlib/v/tests/fns/param_fixed_array_const_test.v
@@ -1,0 +1,26 @@
+module main
+
+import crypto.ed25519
+
+struct SSHClientHello {
+	version    u8
+	public_key [crypto.ed25519.public_key_size]u8
+}
+
+struct SSHServerHello {
+	version u8
+	send_me [crypto.ed25519.public_key_size]u8
+}
+
+struct SSHClientProof {
+	result [crypto.ed25519.public_key_size]u8
+}
+
+fn new_client_hello(public_key [ed25519.public_key_size]u8) {
+	assert public_key.len == ed25519.public_key_size
+	assert public_key.len != 0
+}
+
+fn test_main() {
+	new_client_hello([ed25519.public_key_size]u8{})
+}

